### PR TITLE
[JP Plugin] Install Overlay UI

### LINKED ISSUE #20016: Create the fullscreen modal
Refs qk5s4q4mZsXknwhMWo2vIY-fi-8221%3A20788
Depends on #20015 

Image

## Tasks

- [x] Create the fullscreen modal UI. (#20067)
- [x] Dismiss the modal and show My Sites when the Close button is tapped. #20132 
- [x] Show the Help screen when the Contact Support button is tapped. #20166
- [x] Add Tracks. #20132

### Unresolved issues

From #20067:
- [x] The t&c label collides with the body text in accessibility size. #20165 
- [x] Body text is not yet configured based on single or multiple individual plugins. #20166

From #20132:
- [x] Make the auto-dismiss behavior optional in the `JetpackFullscreenOverlayViewController` when the secondary button is tapped. We need this to show the Support screen instead of having the whole thing dismissed. #20166
- [x] Show the terms & conditions page via the in-app browser when the link is tapped. #20166

### PR DESCRIPTION
**Pull request by alpavanoglu**
Fixes #20016 

This PR only adds the view and has no access point just yet. So testing is only possible with the test code provided below.

To test:
Replace the `presentOverlayIfNeeded` function in `JetpackFeaturesRemovalCoordinator.swift` as follows:

```Swift
    static func presentOverlayIfNeeded(in viewController: UIViewController,
                                       source: OverlaySource,
                                       forced: Bool = false,
                                       fullScreen: Bool = false,
                                       blog: Blog? = nil,
                                       onWillDismiss: JetpackOverlayDismissCallback? = nil,
                                       onDidDismiss: JetpackOverlayDismissCallback? = nil) {
        let phase = generalPhase()
        let frequencyConfig = phase.frequencyConfig
        let frequencyTrackerPhaseString = source.frequencyTrackerPhaseString(phase: phase)
//        var viewModel = JetpackFullscreenOverlayGeneralViewModel(phase: phase, source: source, blog: blog)
        var viewModel = JetpackPluginOverlayViewModel(siteName: "https://test.wordpress.com")
        viewModel.onWillDismiss = onWillDismiss
        viewModel.onDidDismiss = onDidDismiss
        let frequencyTracker = JetpackOverlayFrequencyTracker(frequencyConfig: frequencyConfig,
                                                              phaseString: frequencyTrackerPhaseString,
                                                              source: source)
//        guard viewModel.shouldShowOverlay, frequencyTracker.shouldShow(forced: forced) else {
//            onWillDismiss?()
//            onDidDismiss?()
//            return
//        }
        createAndPresentOverlay(with: viewModel, in: viewController, fullScreen: fullScreen)
        frequencyTracker.track()
    }
```

1. Launch Jetpack app
2. The modal should appear. Verify if the design matches and string attributes look meaningful.
3. Verify if the close button, closes the dialog

## Regression Notes
1. Potential unintended areas of impact
N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackPluginOverlayViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+            %1$@ is using the %2$@, which doesn't support all features of the app yet.
+
+            Please install the %3$@ to use the app with this site.
+            """,
+            comment: """
+            Jetpack Plugin Modal subtitle with formatted texts.
+            One is for the site name, the other for 'Jetpack Backup'
+            and the last one for 'full Jetpack Plugin'
+            """
+        )
```
Comment (dvdchr) [inline WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackPluginOverlayViewModel.swift]: The subtitle needs to be differentiated based on how many individual plugins are installed:

- If the site only has **exactly one** individual plugin: 
  `"{site URL} is using the {plugin name} plugin, which doesn’t …"`.

- But if the site has **more than one** individual plugins, the text should be: 
  `"{site URL} is using individual plugins, which doesn’t …"`

Note that for the `{plugin name}`, you'll also need to manually parse the plugin names based on the value from `jetpack_connection_active_plugins`. I've just added the mapping list to the spec, and you can find them here: pctCYC-An-p2#appendix (appendix 2).

---
Also, it pains me to know that we are just one version away from being able to use [`AttributedString(markdown:options:baseURL:)`](https://developer.apple.com/documentation/foundation/attributedstring/3796160-init) 😅

File: WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackPluginOverlayViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+        static let jetpackBackupText = NSLocalizedString(
+            "jetpack.plugin.modal.subtitle.jetpack.backup",
+            value: "Jetpack Backup",
+            comment: "The 'Jetpack Backup' string in the subtitle"
+        )
```
Comment (dvdchr) [inline WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackPluginOverlayViewModel.swift]: Plugin names shouldn't be localized, as it's akin to a product name. Referring to my previous comment, please head over to pctCYC-An-p2#appendix (appendix 2) to see the full mapping of the possible Jetpack plugin names.

File: WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackFullscreenOverlayViewController.swift [hunk 3]
```diff
@@ -173,6 +176,8 @@ class JetpackFullscreenOverlayViewController: UIViewController {
         subtitleLabel.adjustsFontForContentSizeCategory = true
         footnoteLabel.font = WPStyleGuide.fontForTextStyle(.body, fontWeight: .regular)
         footnoteLabel.adjustsFontForContentSizeCategory = true
+        actionInfoLabel.font = WPStyleGuide.fontForTextStyle(.body, fontWeight: .regular)
+        actionInfoLabel.adjustsFontForContentSizeCategory = true
```
Comment (dvdchr) [inline WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackFullscreenOverlayViewController.swift]: Currently, both the `actionInfoLabel` and the `subtitleLabel` is displayed with the same font size. As per the design, I think the terms and conditions part should use the `subhead` size.

Screenshot 2023-02-08 at 00 26 40

File: WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackFullscreenOverlayViewController.swift [hunk 4]
```diff
@@ -161,6 +163,7 @@ class JetpackFullscreenOverlayViewController: UIViewController {
     private func setupColors() {
         view.backgroundColor = Colors.backgroundColor
         footnoteLabel.textColor = Colors.footnoteTextColor
+        actionInfoLabel.textColor = Colors.footnoteTextColor
```
Comment (dvdchr) [inline WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackFullscreenOverlayViewController.swift]: While looking similar in light mode, the footnote color looks competing with other components in dark mode. As per the design, this component should use a `secondaryLabel` or `UIColor.textSubtle` color (both are the same). Here's a comparison:

Footnote color | textSubtle color
-|-
footnote_color | textsubtle_color

File: WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackPluginOverlayViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+        return NSAttributedString(
+            format: attributedSubtitle(
+                with: Strings.subtitle,
+                fontWeight: .regular),
+            args: ("%1$@", siteNameAttributedText), ("%2$@", jetpackBackupAttributedText), ("%3$@", jetpackPluginAttributedText)
+        )
+    }
+
+    private static func footnote() -> NSAttributedString {
```
Comment (dvdchr) [inline WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackPluginOverlayViewModel.swift]: Hmm, should this be `actionInfoString` instead? It's a bit confusing to me that we have both `footnote` and `actionInfoText`, but this function is the one that gets assigned to `actionInfoText` 😅 .

File: WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackPluginOverlayViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                .font: footnoteBaseFont,
+                .underlineStyle: NSUnderlineStyle.single.rawValue
+            ]
+        )
+
+        let font = WPStyleGuide.fontForTextStyle(.body, fontWeight: .regular)
+        return NSAttributedString(
+            format: footnoteBaseText,
+            args: ("%@", footnoteTermsText)
+        )
```
Comment (dvdchr) [inline WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackPluginOverlayViewModel.swift]: **Optional** 
Just FYI, we already have a convenient extension [String+Underline](https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/blob/trunk/WordPressAuthenticator/Extensions/String%2BUnderline.swift) that handles this scenario. Although, you'll have to supply an underscore to the localized string to denote which part gets to be underlined:

```swift
/// Creates an attributed string from one underlined section that's surrounded by underscores
///
/// - Parameters:
///   - color: foreground color to use for the string (optional)
///   - underlineColor: foreground color to use for the underlined section (optional)
/// - Returns: Attributed string
/// - Note: "this _is_ underlined" would under the "is"
func underlined(color: UIColor? = nil, underlineColor: UIColor? = nil) -> NSAttributedString {
```

**Review by dvdchr — CHANGES_REQUESTED** [review]
@alpavanoglu 👋🏼 ! Thanks for working on this.

When using an accessibility size, the actionInfoLabel collides with the subtitle on AX1 size. Here's a preview:



Perhaps the scroll view's bottom should be constrained to the action info label? This would further minimize the viewport on larger sizes, though. But I'm pretty sure the terms and conditions text needs to remain visible as long as the install button is visible.

I've also tested this on iPad, and apart from the existing issues I've raised, it looks great!

File: WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackFullscreenOverlaySiteCreationViewModel.swift [hunk 7]
```diff
@@ -6,6 +6,9 @@ struct JetpackFullscreenOverlaySiteCreationViewModel: JetpackFullscreenOverlayVi
 
     let phase: JetpackFeaturesRemovalCoordinator.SiteCreationPhase
     let source: String
+    let actionInfoText: NSAttributedString? = nil
+    let footnote: String? = nil
+    let learnMoreButtonURL: String? = nil
```
Comment (dvdchr) [inline WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackFullscreenOverlaySiteCreationViewModel.swift:11]: I'm not sure which one this button is. Is this for the terms & conditions? Assuming this is still in progress, perhaps this will be clearer in your next PRs.

File: WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackPluginOverlayViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+        static let lottieRTLFileName = "JetpackInstallPluginLogoAnimation_rtl"
+    }
+
+    let title: String = Strings.title
+    let subtitle: NSAttributedString
+    let animationLtr: String = Constants.lottieLTRFileName
+    let animationRtl: String = Constants.lottieRTLFileName
+    let footnote: String? = nil
+    let actionInfoText: NSAttributedString? = JetpackPluginOverlayViewModel.footnote()
+    let learnMoreButtonURL: String? = nil
```
Comment (dvdchr) [inline WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackPluginOverlayViewModel.swift:20]: I'm not sure which button this is. Is this for the terms & conditions URL? 🤔 Seeing that this is still in progress, I guess this would be clearer in your next PRs.

Comment (alpavanoglu) [reply WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackPluginOverlayViewModel.swift:20]: Yeah I renamed many things locally. I can update it along with those there where the name will make more sense

File: WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackFullscreenOverlaySiteCreationViewModel.swift (hunk 7 above)
Comment (dvdchr) [inline WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackFullscreenOverlaySiteCreationViewModel.swift:11]: Is this change on this file intended?

File: WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackFullscreenOverlaySiteCreationViewModel.swift [hunk 9]
```diff
@@ -41,14 +44,6 @@ struct JetpackFullscreenOverlaySiteCreationViewModel: JetpackFullscreenOverlayVi
...
     }
 
-    var footnote: String? {
-        return nil
-    }
-
-    var learnMoreButtonURL: String? {
-        return nil
-    }
-
```
Comment (dvdchr) [inline WordPress/Classes/ViewRelated/Jetpack/Branding/Fullscreen Overlay/JetpackFullscreenOverlaySiteCreationViewModel.swift:51]: Same as the above, as it looks like this is a change on an unrelated file. But I guess it's a harmless change that results in a lesser code — there's nothing to compute if it always returns `nil`.

**Review by dvdchr — APPROVED** [review]
LGTM (with some notes), as this is a WIP.

Some issues that weren't addressed yet, but will be addressed later:

- Unused variables or semantically incorrect var names in `JetpackPluginOverlayViewModel`,
- Unresolved AX1 display issues for the overlay

Thanks, @alpavanoglu !
